<?xml version="1.0"?>

<robot xmlns:xacro="http://www.ros.org/wiki/xacro">
  <gazebo>
    <plugin name="motor_control_plugin" filename="libmotor_control_plugin.so"/>
  </gazebo>

  <xacro:property name="pi" value="3.1415926535897931"/>
  
  <xacro:property name="size_factor" value="1.0"/>
  
  <xacro:property name="base_radius" value="${size_factor/8}"/>
  
  <xacro:property name="wheel_axel_length" value="${size_factor}"/>
  <xacro:property name="wheel_axel_rad" value="0.01"/>

  <xacro:property name="motor_axel_length" value="${size_factor}"/>
  <xacro:property name="motor_axel_rad" value="0.01"/>

  <xacro:property name="motor_rad" value="${size_factor/8}"/>
  <xacro:property name="motor_h" value="0.005"/>

  <xacro:property name="wheel_rad" value="${size_factor/2}"/>
  <xacro:property name="wheel_h" value="0.05"/>

  <xacro:property name="motor_connector_rad" value="0.005"/>
  <xacro:property name="motor_connector_h" value="0.05"/>

  <xacro:property name="base_mass" value="0.5"/>
  <xacro:property name="base_inertia" value="${2*base_mass*base_radius*base_radius/5}"/>
  
  <xacro:property name="wheel_axel_mass" value="0.1"/>
  <xacro:property name="wheel_axel_in_ixx" value="${(wheel_axel_mass/12)*(3*wheel_axel_rad*wheel_axel_rad + wheel_axel_length*wheel_axel_length)}"/>
  <xacro:property name="wheel_axel_in_izz" value="${wheel_axel_mass*wheel_axel_rad*wheel_axel_rad/2}"/>

  <xacro:property name="motor_axel_mass" value="0.1"/>
  <xacro:property name="motor_axel_in_ixx" value="${(motor_axel_mass/12)*(3*motor_axel_rad*motor_axel_rad + motor_axel_length*motor_axel_length)}"/>
  <xacro:property name="motor_axel_in_izz" value="${motor_axel_mass*motor_axel_rad*motor_axel_rad/2}"/>

  <xacro:property name="motor_mass" value="0.2"/>
  <xacro:property name="motor_ixx" value="${(motor_mass/12)*(3*motor_rad*motor_rad + motor_h*motor_h)}"/>
  <xacro:property name="motor_izz" value="${motor_mass*motor_rad*motor_rad/2}"/>
  
  <xacro:property name="wheel_mass" value="0.5"/>
  <xacro:property name="wheel_ixx" value="${(wheel_mass/12)*(6*wheel_rad*wheel_rad + wheel_h*wheel_h)}"/>
  <xacro:property name="wheel_izz" value="${wheel_mass*wheel_rad*wheel_rad}"/>

  <xacro:property name="motor_connector_mass" value="0.01"/>
  <xacro:property name="motor_connector_ixx" value="${(motor_connector_mass/12)*(3*motor_connector_rad*motor_connector_rad + motor_connector_h*motor_connector_h)}"/>
  <xacro:property name="motor_connector_izz" value="${motor_connector_mass*motor_connector_rad*motor_connector_rad/2}"/>

  <!-- Main Quadrotor Link -->
  <xacro:macro name="quadrotor_base_macro">

    <!-- Links Start -->

    <!-- Base Link -->
    <link name="base_link">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${base_radius}"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <sphere radius="${base_radius}"/>
        </geometry>
      </collision>
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="${base_mass}"/>
        <inertia
          ixx="${base_inertia}" ixy="0.0" ixz="0.0"
          iyy="${base_inertia}" iyz="0.0"
          izz="${base_inertia}"
        />
      </inertial>
    </link>
    <gazebo reference="base_link">
      <material>Gazebo/Blue</material>
    </gazebo>

    <!-- Axis to wheels -->
    <link name="wheel_axel">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${wheel_axel_rad}" length="${wheel_axel_length}"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${wheel_axel_rad}" length="${wheel_axel_length}"/>
        </geometry>
      </collision>
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="${wheel_axel_mass}"/>
        <inertia
          ixx="${wheel_axel_in_ixx}" ixy="0.0" ixz="0.0"
          iyy="${wheel_axel_in_ixx}" iyz="0.0"
          izz="${wheel_axel_in_izz}"
        />
      </inertial>
    </link>
    <gazebo reference="wheel_axel">
      <material>Gazebo/Black</material>
    </gazebo>

    <!-- Axis 1 to rotors -->
    <link name="motor_axel_1">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${motor_axel_rad}" length="${motor_axel_length}"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${motor_axel_rad}" length="${motor_axel_length}"/>
        </geometry>
      </collision>
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="${motor_axel_mass}"/>
        <inertia
          ixx="${motor_axel_in_ixx}" ixy="0.0" ixz="0.0"
          iyy="${motor_axel_in_ixx}" iyz="0.0"
          izz="${motor_axel_in_izz}"
        />
      </inertial>
    </link>
    <gazebo reference="motor_axel_1">
      <material>Gazebo/Black</material>
    </gazebo>

    <!-- Axis 2 to rotors -->
    <link name="motor_axel_2">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${motor_axel_rad}" length="${motor_axel_length}"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${motor_axel_rad}" length="${motor_axel_length}"/>
        </geometry>
      </collision>
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="${motor_axel_mass}"/>
        <inertia
          ixx="${motor_axel_in_ixx}" ixy="0.0" ixz="0.0"
          iyy="${motor_axel_in_ixx}" iyz="0.0"
          izz="${motor_axel_in_izz}"
        />
      </inertial>
    </link>
    <gazebo reference="motor_axel_2">
      <material>Gazebo/Black</material>
    </gazebo>

    <!-- Motor 1 -->
    <link name="motor1">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${motor_rad}" length="${motor_h}"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${motor_rad}" length="${motor_h}"/>
        </geometry>
      </collision>
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="${motor_mass}"/>
        <inertia
          ixx="${motor_ixx}" ixy="0.0" ixz="0.0"
          iyy="${motor_ixx}" iyz="0.0"
          izz="${motor_izz}"
        />
      </inertial>
    </link>
    <gazebo reference="motor1">
      <material>Gazebo/Red</material>
    </gazebo>

    <!-- Motor 2 -->
    <link name="motor2">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${motor_rad}" length="${motor_h}"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${motor_rad}" length="${motor_h}"/>
        </geometry>
      </collision>
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="${motor_mass}"/>
        <inertia
          ixx="${motor_ixx}" ixy="0.0" ixz="0.0"
          iyy="${motor_ixx}" iyz="0.0"
          izz="${motor_izz}"
        />
      </inertial>
    </link>
    <gazebo reference="motor2">
      <material>Gazebo/Red</material>
    </gazebo>

    <!-- Motor 3 -->
    <link name="motor3">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${motor_rad}" length="${motor_h}"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${motor_rad}" length="${motor_h}"/>
        </geometry>
      </collision>
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="${motor_mass}"/>
        <inertia
          ixx="${motor_ixx}" ixy="0.0" ixz="0.0"
          iyy="${motor_ixx}" iyz="0.0"
          izz="${motor_izz}"
        />
      </inertial>
    </link>
    <gazebo reference="motor3">
      <material>Gazebo/Green</material>
    </gazebo>

    <!-- Motor 4 -->
    <link name="motor4">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${motor_rad}" length="${motor_h}"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${motor_rad}" length="${motor_h}"/>
        </geometry>
      </collision>
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="${motor_mass}"/>
        <inertia
          ixx="${motor_ixx}" ixy="0.0" ixz="0.0"
          iyy="${motor_ixx}" iyz="0.0"
          izz="${motor_izz}"
        />
      </inertial>
    </link>
    <gazebo reference="motor4">
      <material>Gazebo/Green</material>
    </gazebo>

    <!-- Wheel 1 -->
    <link name="wheel1">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${wheel_rad}" length="${wheel_h}"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${wheel_rad}" length="${wheel_h}"/>
        </geometry>
      </collision>
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <mass value="${wheel_mass}"/>
        <inertia
          ixx="${wheel_ixx}" ixy="0.0" ixz="0.0"
          iyy="${wheel_ixx}" iyz="0.0"
          izz="${wheel_izz}"
        />
      </inertial>
    </link>
    <gazebo reference="wheel1">
      <material>Gazebo/White</material>
    </gazebo>

    <!-- Wheel 2 -->
    <link name="wheel2">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${wheel_rad}" length="${wheel_h}"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="${wheel_rad}" length="${wheel_h}"/>
        </geometry>
      </collision>
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <mass value="${wheel_mass}"/>
        <inertia
          ixx="${wheel_ixx}" ixy="0.0" ixz="0.0"
          iyy="${wheel_ixx}" iyz="0.0"
          izz="${wheel_izz}"
        />
      </inertial>
    </link>
    <gazebo reference="wheel2">
      <material>Gazebo/White</material>
    </gazebo>

    <!-- Motor Connector 1 -->
    <link name="mc1">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
          <cylinder radius="${motor_connector_rad}" length="${motor_connector_h}"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
          <cylinder radius="${motor_connector_rad}" length="${motor_connector_h}"/>
        </geometry>
      </collision>
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <mass value="${motor_connector_mass}"/>
        <inertia
          ixx="${motor_connector_ixx}" ixy="0.0" ixz="0.0"
          iyy="${motor_connector_ixx}" iyz="0.0"
          izz="${motor_connector_izz}"
        />
      </inertial>
    </link>
    <gazebo reference="mc1">
      <material>Gazebo/Black</material>
    </gazebo>

    <!-- Motor Connector 2 -->
    <link name="mc2">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
          <cylinder radius="${motor_connector_rad}" length="${motor_connector_h}"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
          <cylinder radius="${motor_connector_rad}" length="${motor_connector_h}"/>
        </geometry>
      </collision>
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <mass value="${motor_connector_mass}"/>
        <inertia
          ixx="${motor_connector_ixx}" ixy="0.0" ixz="0.0"
          iyy="${motor_connector_ixx}" iyz="0.0"
          izz="${motor_connector_izz}"
        />
      </inertial>
    </link>
    <gazebo reference="mc2">
      <material>Gazebo/Black</material>
    </gazebo>

    <!-- Motor Connector 3 -->
    <link name="mc3">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
          <cylinder radius="${motor_connector_rad}" length="${motor_connector_h}"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
          <cylinder radius="${motor_connector_rad}" length="${motor_connector_h}"/>
        </geometry>
      </collision>
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <mass value="${motor_connector_mass}"/>
        <inertia
          ixx="${motor_connector_ixx}" ixy="0.0" ixz="0.0"
          iyy="${motor_connector_ixx}" iyz="0.0"
          izz="${motor_connector_izz}"
        />
      </inertial>
    </link>
    <gazebo reference="mc3">
      <material>Gazebo/Black</material>
    </gazebo>

    <!-- Motor Connector 4 -->
    <link name="mc4">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
          <cylinder radius="${motor_connector_rad}" length="${motor_connector_h}"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <geometry>
          <cylinder radius="${motor_connector_rad}" length="${motor_connector_h}"/>
        </geometry>
      </collision>
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <mass value="${motor_connector_mass}"/>
        <inertia
          ixx="${motor_connector_ixx}" ixy="0.0" ixz="0.0"
          iyy="${motor_connector_ixx}" iyz="0.0"
          izz="${motor_connector_izz}"
        />
      </inertial>
    </link>
    <gazebo reference="mc4">
      <material>Gazebo/Black</material>
    </gazebo>

    <!-- Links End -->

    <!-- Joints Start-->

    <!-- Base link to wheel axel -->
    <joint name="base_wheel_axel_joint" type="fixed">
      <origin xyz="0 0 ${base_radius/2}" rpy="${-pi/2} 0 0"/>
      <parent link="base_link"/>
      <child link="wheel_axel"/>
    </joint>

    <!-- Base link to motor axel 1 -->
    <joint name="base_motor_axel1_joint" type="fixed">
      <origin xyz="0 0 ${base_radius/2}" rpy="${-pi/2} 0 ${-pi/4}"/>
      <parent link="base_link"/>
      <child link="motor_axel_1"/>
    </joint>

    <!-- Base link to motor axel 2 -->
    <joint name="base_motor_axel2_joint" type="fixed">
      <origin xyz="0 0 ${base_radius/2}" rpy="${pi/2} 0 ${pi/4}"/>
      <parent link="base_link"/>
      <child link="motor_axel_2"/>
    </joint>

    <!-- Wheel axel to wheel 1 -->
    <joint name="axel_wheel1_joint" type="continuous">
      <origin xyz="0 0 ${wheel_axel_length/2}" rpy="0 0 0"/>
      <parent link="wheel_axel"/>
      <child link="wheel1"/>
      <axis xyz="0 0 1"/>
      <dynamics damping="0"/>
    </joint>

    <!-- Wheel axel to wheel 2 -->
    <joint name="axel_wheel2_joint" type="continuous">
      <origin xyz="0 0 ${-wheel_axel_length/2}" rpy="0 0 0"/>
      <parent link="wheel_axel"/>
      <child link="wheel2"/>
      <axis xyz="0 0 1"/>
      <dynamics damping="0"/>
    </joint>

    <!-- Motor axel 1 to motor con 1 -->
    <joint name="axel1_motor_con1_joint" type="fixed">
      <origin xyz="0 0 ${(motor_axel_length/2)-motor_rad}" rpy="${pi/2} 0 0"/>
      <parent link="motor_axel_1"/>
      <child link="mc1"/>
    </joint>

    <!-- Motor axel 1 to motor con 3 -->
    <joint name="axel1_motor_con3_joint" type="fixed">
      <origin xyz="0 0 ${motor_rad-(motor_axel_length/2)}" rpy="${pi/2} 0 0"/>
      <parent link="motor_axel_1"/>
      <child link="mc3"/>
    </joint>

    <!-- Motor axel 2 to motor con 2 -->
    <joint name="axel2_motor_con2_joint" type="fixed">
      <origin xyz="0 0 ${(motor_axel_length/2)-motor_rad}" rpy="${-pi/2} 0 0"/>
      <parent link="motor_axel_2"/>
      <child link="mc2"/>
    </joint>

    <!-- Motor axel 2 to motor con 4 -->
    <joint name="axel2_motor_con4_joint" type="fixed">
      <origin xyz="0 0 ${motor_rad-(motor_axel_length/2)}" rpy="${-pi/2} 0 0"/>
      <parent link="motor_axel_2"/>
      <child link="mc4"/>
    </joint>

    <!-- Motor con 1 to motor 1 -->
    <joint name="mc1_motor1_joint" type="revolute">
      <origin xyz="0 0 ${motor_connector_h/2}" rpy="0 0 0"/>
      <parent link="mc1"/>
      <child link="motor1"/>
      <limit effort="1000.0" velocity="0.0" lower="0.0" upper="0.0"/>
    </joint>

    <!-- Motor con 2 to motor 2 -->
    <joint name="mc2_motor2_joint" type="revolute">
      <origin xyz="0 0 ${motor_connector_h/2}" rpy="0 0 0"/>
      <parent link="mc2"/>
      <child link="motor2"/>
      <limit effort="1000.0" velocity="0.0" lower="0.0" upper="0.0"/>
    </joint>

    <!-- Motor con 3 to motor 3 -->
    <joint name="mc3_motor3_joint" type="revolute">
      <origin xyz="0 0 ${motor_connector_h/2}" rpy="0 0 0"/>
      <parent link="mc3"/>
      <child link="motor3"/>
      <limit effort="1000.0" velocity="0.0" lower="0.0" upper="0.0"/>
    </joint>

    <!-- Motor con 4 to motor 4 -->
    <joint name="mc4_motor4_joint" type="revolute">
      <origin xyz="0 0 ${motor_connector_h/2}" rpy="0 0 0"/>
      <parent link="mc4"/>
      <child link="motor4"/>
      <limit effort="1000.0" velocity="0.0" lower="0.0" upper="0.0"/>
    </joint>

    <!-- Joints End -->

  </xacro:macro>

</robot>
